compiler/cpp/src/generate/t_java_generator.cc

```diff
@@ -3967,14 +3972,42 @@ string t_java_generator::function_signature_async(t_function* tfunction, bool us
  * @param tfunction Function definition
  * @return String of rendered function definition
  */
-string t_java_generator::function_signature_future(t_function* tfunction, string prefix) {
+string t_java_generator::function_signature_future(t_function* tfunction, int params_futures_mask, string prefix) {
   t_type* ttype = tfunction->get_returntype();
   std::string return_type = type_name(ttype, true);
   std::string fn_name = get_rpc_method_name(tfunction->get_name());
-  std::string arg_list = argument_list(tfunction->get_arglist());
+  std::string arg_list = futures_argument_list(tfunction->get_arglist(), params_futures_mask);
   return "CompletableFuture<" + return_type + "> " + prefix + fn_name + "(" + arg_list + ")";
 }
 
+/**
+ * Renders a comma separated field list, with type names. The types are
+ * modified to use a future of the argument type according to the mask
+ * given.
+ */
+string t_java_generator::futures_argument_list(t_struct* tstruct, int params_futures_mask) {
+  string result = "";
+
+  const vector<t_field*>& fields = tstruct->get_members();
+  vector<t_field*>::const_iterator f_iter;
+  int i = 0;
+  for (f_iter = fields.begin(); f_iter != fields.end(); ++f_iter, ++i) {
+    if (i > 0) {
+      result += ", ";
+    }
+    string type;
+    if (params_futures_mask & (1 << i)) {
+      // The mask indicates to use a future for this argument
+      type = "CompletableFuture<" + type_name((*f_iter)->get_type(), true) + ">";
```

Comment:
Ok, I now think I understood the problem. You were probably talking about something receiving Future< Long > and feeding it a Future< Integer > ? As you said, better avoid this now. We could later provide converting functions for number type promotion futures Futures.toLong(Future< Integer > intFuture); It's not ideal but I don't think it will be a super common use case.